exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/join/HashJoinProbeTemplate.java

```diff
@@ -98,7 +102,9 @@ public void setupHashJoinProbe(FragmentContext context, VectorContainer buildBat
   }
 
   public void executeProjectRightPhase() {
-    while (outputRecords < TARGET_RECORDS_PER_BATCH && recordsProcessed < recordsToProcess) {
+    while (outputRecords < targetRecordsPerBatch
+            && recordsProcessed < recordsToProcess
+            && (!adjustTargetRecordsPerBatch || outgoingJoinBatch.getMemoryUsed() < TARGET_BATCH_SIZE_IN_BYTES)) {
```

Comment:
It seems like the thing we are testing for here isn't actually directly related to the condition we are trying to avoid. The overall memory consumed when outputting records will be a function of both size of values as well as number of columns. I think this is a reasonable approach for now but we should open a follow-up JIRA to look at where things will break as we encounter cases where there are many wide columns in a dataset.